hacks/deco.man
.TH XScreenSaver 1 "27-Apr-97" "X Version 11"
.SH NAME
deco - draw tacky 70s basement wall panelling
.SH SYNOPSIS
.B deco
[\-display \fIhost:display.screen\fP] [\-foreground \fIcolor\fP] [\-background \fIcolor\fP] [\-window] [\-root] [\-mono] [\-install] [\-visual \fIvisual\fP] [\-delay \fIseconds\fP] [\-max\-depth \fIint\fP] [\-min\-width \fIint\fP] [\-min\-height \fIint\fP]
.SH DESCRIPTION
The \fIdeco\fP program subdivides and colors rectangles randomly.
It looks kind of like Brady-Bunch-era rec-room wall paneling.
(Raven says: "this screensaver is ugly enough to peel paint.")
.SH OPTIONS
.I deco
accepts the following options:
.TP 8
.B \-window
Draw on a newly-created window.  This is the default.
.TP 8
.B \-root
Draw on the root window.
.TP 8
.B \-mono 
If on a color display, pretend we're on a monochrome display.
.TP 8
.B \-install
Install a private colormap for the window.
.TP 8
.B \-visual \fIvisual\fP
Specify which visual to use.  Legal values are the name of a visual class,
or the id number (decimal or hex) of a specific visual.
.TP 8
.B \-delay \fIseconds\fP
How long to wait before starting over.  Default 5 seconds.
.TP 8
.B \-max\-depth \fIinteger\fP
How deep to subdivide.  Default 12.
Default 8.
.TP 8
.B \-min-width \fIinteger\fP
.B \-min-height \fIinteger\fP
The size of the smallest rectangle to draw.  Default 20x20.
.SH ENVIRONMENT
.PP
.TP 8
.B DISPLAY
to get the default host and display number.
.TP 8
.B XENVIRONMENT
to get the name of a resource file that overrides the global resources
stored in the RESOURCE_MANAGER property.
.SH SEE ALSO
.BR X (1),
.BR xscreensaver (1)
.SH COPYRIGHT
Copyright \(co 1997 by Jamie Zawinski.  Permission to use, copy, modify, 
distribute, and sell this software and its documentation for any purpose is 
hereby granted without fee, provided that the above copyright notice appear 
in all copies and that both that copyright notice and this permission notice
appear in supporting documentation.  No representations are made about the 
suitability of this software for any purpose.  It is provided "as is" without
express or implied warranty.
.SH AUTHOR
Jamie Zawinski <jwz@netscape.com>, 26-Apr-97, based on code by
Michael D. Bayne <mdb@go2net.com>.


hacks/deco.c
/* xscreensaver, Copyright (c) 1997 Jamie Zawinski <jwz@netscape.com>
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided that
 * the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation.  No representations are made about the suitability of this
 * software for any purpose.  It is provided "as is" without express or 
 * implied warranty.
 *
 * Concept snarfed from Michael D. Bayne in
 * http://www.go2net.com/internet/deep/1997/04/16/body.html
 */

#include "screenhack.h"
#include <stdio.h>

static XColor colors[255];
static int ncolors = 0;
static int max_depth = 0;
static int min_height = 0;
static int min_width = 0;

static void
#ifdef __STDC__
deco (Display *dpy,
      Window window,
      Colormap cmap,
      GC fgc, GC bgc,
      int x, int y, int w, int h, int depth)
#else /* !__STDC__ */
deco (dpy, window, cmap, fgc, bgc, x, y, w, h, depth)
  Display *dpy;
  Window window;
  Colormap cmap;
  GC fgc, bgc;
  int x, y, w, h, depth;
#endif /* !__STDC__ */
{
  if (((random() % max_depth) < depth) || (w < min_width) || (h < min_height))
    {
      if (!mono_p)
	{
	  static int current_color = 0;
	  if (++current_color >= ncolors)
	    current_color = 0;
	  XSetForeground(dpy, bgc, colors[current_color].pixel);
	}
      XFillRectangle (dpy, window, bgc, x, y, w, h);
      XDrawRectangle (dpy, window, fgc, x, y, w, h);
    }
  else
    {
      if (random() & 1)
	{
	  deco (dpy, window, cmap, fgc, bgc, x, y, w/2, h, depth+1);
	  deco (dpy, window, cmap, fgc, bgc, x+w/2, y, w/2, h, depth+1);
	}
      else
	{
	  deco (dpy, window, cmap, fgc, bgc, x, y, w, h/2, depth+1);
	  deco (dpy, window, cmap, fgc, bgc, x, y+h/2, w, h/2, depth+1);
	}
    }
}


char *progclass = "Deco";

char *defaults [] = {
  "Deco.background:	black",		/* to placate SGI */
  "Deco.foreground:	white",
  "*maxDepth:		12",
  "*minWidth:		20",
  "*minHeight:		20",
  "*cycle:		False",
  "*delay:		5",
  "*cycleDelay:		1000000",
  "*ncolors:		64",
  0
};

XrmOptionDescRec options [] = {
  { "-max-depth",	".maxDepth",	XrmoptionSepArg, 0 },
  { "-min-width",	".minWidth",	XrmoptionSepArg, 0 },
  { "-min-height",	".minHeight",	XrmoptionSepArg, 0 },
  { "-delay",		".delay",	XrmoptionSepArg, 0 },
  { "-ncolors",		".ncolors",	XrmoptionSepArg, 0 },
  { "-cycle",		".cycle",	XrmoptionNoArg, "True" },
  { "-no-cycle",	".cycle",	XrmoptionNoArg, "False" },
  { "-cycle-delay",	".cycleDelay",	XrmoptionSepArg, 0 },
  { 0, 0, 0, 0 }
};

void
#ifdef __STDC__
screenhack (Display *dpy, Window window)
#else /* ! __STDC__ */
screenhack (dpy, window) Display *dpy; Window window;
#endif /* ! __STDC__ */
{
  GC fgc, bgc;
  XGCValues gcv;
  XWindowAttributes xgwa;
  int delay = get_integer_resource ("delay", "Integer");
  int cycle_delay = get_integer_resource ("cycleDelay", "Integer");
  Bool writable = get_boolean_resource ("cycle", "Boolean");

  max_depth = get_integer_resource ("maxDepth", "Integer");
  if (max_depth < 1) max_depth = 1;
  else if (max_depth > 1000) max_depth = 1000;

  min_width = get_integer_resource ("minWidth", "Integer");
  if (min_width < 2) min_width = 2;
  min_height = get_integer_resource ("minHeight", "Integer");
  if (min_height < 2) min_height = 2;

  XGetWindowAttributes (dpy, window, &xgwa);

  gcv.foreground = get_pixel_resource("foreground", "Foreground",
				      dpy, xgwa.colormap);
  fgc = XCreateGC (dpy, window, GCForeground, &gcv);

  gcv.foreground = get_pixel_resource("background", "Background",
				      dpy, xgwa.colormap);
  bgc = XCreateGC (dpy, window, GCForeground, &gcv);

  ncolors = get_integer_resource ("ncolors", "Integer");

  make_random_colormap (dpy, xgwa.visual, xgwa.colormap, colors, &ncolors,
			False, True, &writable, True);

  if (ncolors <= 2)
    mono_p = True;

  if (!mono_p)
    {
      GC tmp = fgc;
      fgc = bgc;
      bgc = tmp;
    }

  while (1)
    {
      XGetWindowAttributes (dpy, window, &xgwa);
      XFillRectangle(dpy, window, bgc, 0, 0, xgwa.width, xgwa.height);
      deco (dpy, window, xgwa.colormap, fgc, bgc,
	    0, 0, xgwa.width, xgwa.height, 0);
      XSync (dpy, True);

      if (!delay) continue;
      if (!writable)
	sleep (delay);
      else
	{
	  time_t start = time((time_t) 0);
	  while (start - delay < time((time_t) 0))
	    {
	      rotate_colors (dpy, xgwa.colormap, colors, ncolors, 1);
	      if (cycle_delay)
		usleep (cycle_delay);
	    }
	}
    }
}
